### ISSUE TITLE
Modals styling issues when using @mantine/remix

### ISSUE THREAD

**Issue by viczam** [mantinedev/mantine#2311]
### What package has an issue

@mantine/modals

### Describe the bug

Steps to reproduce:
- I started with mantine-remix-template
- I added `@mantine/modals` to package.json
- In root.tsx I added ModalsProvider
```tsx
<ModalsProvider
  modalProps={{ centered: true }}
>
  <Outlet />
</ModalsProvider>
```

- I added this code to index.tsx

```tsx
import { Text, Button, Stack } from '@mantine/core';
import { openConfirmModal } from '@mantine/modals';

export default function Index() {
  const openModal = () =>
    openConfirmModal({
      title: 'Please confirm your action',
      children: (
        <Text size="sm">
          This action is so important that you are required to confirm it with a modal. Please click
          one of these buttons to proceed.
        </Text>
      ),
      labels: { confirm: 'Confirm', cancel: 'Cancel' },
      onCancel: () => console.log('Cancel'),
      onConfirm: () => console.log('Confirmed'),
    });

  return (
    <Stack align="center" mt={50}>
      <Text size="xl" weight={500}>
        Welcome to Mantine!
      </Text>
      <Button onClick={openModal}>Click the button</Button>
    </Stack>
  );
}
```

While the modal looks good in dev mode, after a `yarn build` + `yarn start` the styles are off:
<img width="1728" alt="Screenshot 2022-09-01 at 13 54 11" src="https://user-images.githubusercontent.com/22033/187897642-0d600bfb-424c-466a-a5a9-41dcd268ce57.png">

### What version of @mantine/hooks page do you have in package.json?

5.1.0

### Do you know how to fix the issue

No

### Are you willing to participate in fixing this issue and create a pull request with the fix

Yes

**Comment by rtivital** [comment]
Seems to be an issue with Portal, no idea how to fix that

**Comment by viczam** [comment]
it seems to be related to this - https://github.com/mantinedev/mantine/issues/1734#issuecomment-1208272101

**Comment by rtivital** [comment]
Hi @viczam, I've updated the remix template and guide to support react 18, can you please check whether your issue can be reproduced there?

**Comment by Beardless** [comment]
Hi, @rtivital I have a problem with React 18 and Remix. 

Trying to use your example of integrating Mantine into Remix, but Remix is using now `renderToPipeableStream` instead of `renderToString`. And I don't know where to put `injectStyles` method as I don't have a markup to use it.

This is an example code you will get after initialization of the new remix app `entry.server.tsx`:
``` js
export default function handleRequest(
  request: Request,
  responseStatusCode: number,
  responseHeaders: Headers,
  remixContext: EntryContext
) {
  const callbackName = isbot(request.headers.get("user-agent"))
    ? "onAllReady"
    : "onShellReady";

  return new Promise((resolve, reject) => {
    let didError = false;

    const { pipe, abort } = renderToPipeableStream(
      <RemixServer context={remixContext} url={request.url} />,
      {
        [callbackName]: () => {
          const body = new PassThrough();

          responseHeaders.set("Content-Type", "text/html");

          resolve(
            new Response(body, {
              headers: responseHeaders,
              status: didError ? 500 : responseStatusCode,
            })
          );

          pipe(body);
        },
        onShellError: (err: unknown) => {
          reject(err);
        },
        onError: (error: unknown) => {
          didError = true;

          console.error(error);
        },
      }
    );

    setTimeout(abort, ABORT_DELAY);
  });
}
```

**Comment by rtivital** [comment]
@Beardless it is not related to the issue, template works correctly with react 18. Emotion does not support server streaming, see – https://github.com/emotion-js/emotion/issues/2800

**Comment by stephen776** [comment]
I've updated everything to the latest template and I am still experiencing the modal bug in production builds

**Comment by fatehwaharp** [comment]
When run `npm run build` && 'npm run start' or in production, seems like the emotion styling wont applied to the portal, need to use inline style to fix the modal positioniong

**Comment by fatehwaharp** [comment]
When run `npm run build` && `npm run start` or in production, seems like the emotion styling does not applied in modal's portal.

![CleanShot 2022-09-25 at 23 32 07@2x](https://user-images.githubusercontent.com/306898/192151868-1d2d14f2-15db-4d70-8ee5-b451f5b8b88a.png)

**Comment by stephen776** [comment]
Interestingly... removing the `<ClientProvider> ... </ClientProvider>` from `entry.client.tsx` fixes this issue for me in production.

I am wondering if anyone has any insight on this? Or whether this is a safe change?

**Comment by fatehwaharp** [comment]
> Interestingly... removing the `<ClientProvider> ... </ClientProvider>` from `entry.client.tsx` fixes this issue for me in production.
> 
> I am wondering if anyone has any insight on this? Or whether this is a safe change?

You mean from this:
```typescript
hydrateRoot(
  document,
  <ClientProvider>
    <RemixBrowser />
  </ClientProvider>
)
```
to this ?:

```typescript
hydrateRoot(document);
```

**Comment by stephen776** [comment]
> 

No - keep the `<RemixBrowser />`

**Comment by alukach** [comment]
For the record, removing the `ClientProvider` brings up the following hydration bugs for me:

* https://reactjs.org/docs/error-decoder.html?invariant=418
* https://reactjs.org/docs/error-decoder.html?invariant=423

Reproducible example found here: https://github.com/alukach/mantine-remix-rendering-issue

**Comment by gigobyte** [comment]
I managed to temporarily solve the problem by wrapping my modals in [React.lazy](https://reactjs.org/docs/code-splitting.html#reactlazy) and dynamically importing them. They open dynamically  anyway so this doesn't affect SSR. Depending on your case this might even be beneficial for performance.

**Comment by affanshahid** [comment]
I just ended up copying the non-dynamic styles into a global css file, doesn't work perfectly but is a bearable hack for now. Hoping this gets solved soon.

**Comment by shyamlohar** [comment]
I am facing a similar issue. Happens only on a production build for me as well. did anyone manage to find any workaround for it? I would also love to help if someone can guide me on where i should look to get started

**Comment by aniravi24** [comment]
> Interestingly... removing the `<ClientProvider> ... </ClientProvider>` from `entry.client.tsx` fixes this issue for me in production.
> 
> I am wondering if anyone has any insight on this? Or whether this is a safe change?

I'm noticing the same behavior as this comment - if `<ClientProvider />` is removed, then the modal renders correctly. However, this is not the right solution as ClientProvider is necessarily to properly maintain styles in other situations.

**Comment by smaven** [comment]
any update on this? This is a production critical bug for me.

**Comment by fatehwaharp** [comment]
> any update on this? This is a production critical bug for me.

Have you tried removing the `<ClientProvider>...</ClientProvider>` ? After remove that line, it works for me on production

**Comment by boringd3v** [comment]
`hydrateRoot(
  <RemixBrowser />,
  document
)`

^ Removing the `ClientProvider` works like a charm for me. Thanks.

**Comment by wancup** [comment]
FYI.
I change the emotion cache configuration, and it works fine for my production environment. However, I don't know if this is the correct way to solve this issue. 

```typescript
// entry.client.tsx

import createCache from "@emotion/cache";
import { ClientProvider } from "@mantine/remix";
import { RemixBrowser } from "@remix-run/react";
import { hydrate } from "react-dom";

// to disable prepend option
const cache = createCache({ key: "mantine" });

hydrate(
  <ClientProvider emotionCache={cache}>
    <RemixBrowser />
  </ClientProvider>,
  document
);
```

**Comment by jfeaver** [comment]
I'm running into this issue. I found that removing `ClientProvider` means that you no longer get to take advantage of Emotion caching so I found a different workaround: Hard coding modal styles! If you want to do the same, here's what I did:

```diff
diff --git a/app/integrations/mantine/styles.css b/app/integrations/mantine/styles.css
new file mode 100644
index 0000000..1186223
--- /dev/null
+++ b/app/integrations/mantine/styles.css
@@ -0,0 +1,21 @@
+.mantine-Modal-inner {
+  position: absolute;
+  top: 0;
+  left: 0;
+  right: 0;
+  bottom: 0;
+  overflow-y: auto;
+  padding: 48px 16px;
+  display: -webkit-box;
+  display: -webkit-flex;
+  display: -ms-flexbox;
+  display: flex;
+  -webkit-box-pack: center;
+  -ms-flex-pack: center;
+  -webkit-justify-content: center;
+  justify-content: center;
+  -webkit-align-items: flex-start;
+  -webkit-box-align: flex-start;
+  -ms-flex-align: flex-start;
+  align-items: flex-start;
+}
diff --git a/app/routes/dashboard.tsx b/app/routes/dashboard.tsx
index 6de7e57..68c4ab6 100644
--- a/app/routes/dashboard.tsx
+++ b/app/routes/dashboard.tsx
@@ -1,8 +1,13 @@
-import type { LoaderFunction } from "@remix-run/node";
+import type { LinksFunction, LoaderFunction } from "@remix-run/node";
 import { Outlet, useLoaderData } from "@remix-run/react";
+import styles from "app/integrations/mantine/styles.css";
 import { getSession } from "~/session.server";

+export const links: LinksFunction = () => {
+  return [{ rel: "stylesheet", href: styles }];
+};
+
 export const loader: LoaderFunction = async ({ request }) => {
   const session = await getSession(request);
   const user = session.get("user");
```

**Comment by gregermendle** [comment]
Not sure if this was posted / updated elsewhere, but i was able to get rid of the modal issue by doing the following:
```
let emotionCache = createEmotionCache({ key: "mantine" });

hydrateRoot(
    document,
    <ClientProvider emotionCache={emotionCache}>
        <RemixBrowser />
    </ClientProvider>
);
```

**Comment by affanshahid** [comment]
> Not sure if this was posted / updated elsewhere, but i was able to get rid of the modal issue by doing the following:
> 
> ```js
> let emotionCache = createEmotionCache({ key: "mantine" });
> 
> hydrateRoot(
>     document,
>     <ClientProvider emotionCache={emotionCache}>
>         <RemixBrowser />
>     </ClientProvider>
> );
> ```

This seems to work for me, don't see hydration related errors in the console either. Are there any downsides to this approach? (i.e moving the createEmotionCache from `root.tsx` to `entry.client.tsx`)

**Comment by cayter** [comment]
> Not sure if this was posted / updated elsewhere, but i was able to get rid of the modal issue by doing the following:
> 
> ```js
> let emotionCache = createEmotionCache({ key: "mantine" });
> 
> hydrateRoot(
>     document,
>     <ClientProvider emotionCache={emotionCache}>
>         <RemixBrowser />
>     </ClientProvider>
> );
> ```

Was wondering why this is the case, just realised emotion cache is creating a copy of CSS sheets based on what it can parse. When using Remix, we have to go through SSR on the server-side but components like modal rely on `window` which server-side with NodeJS env (entry.server.tsx) wouldn't have that.

Putting the emotion cache into `entry.client.tsx` works 'cause it's only getting triggered when the Remix codebase starts running on browser side. The downside would be there's now overhead to initiate the emotion cache, though i'm not sure if the emotion cache initialisation will run every time or only when the cache `key` is changed.

**Comment by rtivital** [comment]
This issue is closed for one of these reasons:

- It was fixed in 7.0
- It is no longer reproducible in 7.0
- It is no longer applicable in 7.0
- It does not have a reproduction

If you think that this issue was closed by mistake and it is still an issue in version 7.0, please reopen it.
